packages/driver/src/cypress/mocha.ts

```diff
@@ -416,6 +416,13 @@ const patchRunnableResetTimeout = () => {
     }
 
     this.timer = setTimeout(() => {
+      if (runnable.state === 'passed') {
+        // this timeout can be reached at the same time that a
+        // user does an asynchronous `done`, so double-check
+        // that the test has not already passed before timing out
+        return
+      }
```

Comment:
For anyone curious to reproduce this issue, you can comment this out and runt his test to consistently see the error shown in the video Zach linked: 

```js
      it('should not be true', { defaultCommandTimeout: 1 }, (done) => {
        cy.on('fail', (err) => {
          expect(err.message).to.eq('expected false to be true')

          done()
        })

        cy.noop(false).should('be.true')
      })
```